### ISSUE TITLE
Type Issue

### ISSUE THREAD

**Issue by MelbourneDeveloper** [MelbourneDeveloper/ioc_container#1]
Check out commit: bc5cee026bb91bf37ba481ac55bc1d39bd041007

Run the tests and notice that there is an error. This is the test that fails
![image](https://user-images.githubusercontent.com/16697547/192120119-bc1a9e36-8b83-4974-a8e6-294ad9f4ee61.png)

This is the error that breaks
![image](https://user-images.githubusercontent.com/16697547/192119112-4e90f25a-2721-4309-bc5e-c320c30f78f7.png)

This seems to be a very strange Dart typing issue. For now, working around this by wrapping the `dispose` function to cast to dynamic

**Comment by MelbourneDeveloper** [comment]
This is the current workaround. The aim is to stop using dynamic here. It would be much nicer if we could use `T` like the left version so there is no casting
![image](https://user-images.githubusercontent.com/16697547/192120158-497be127-08b6-40ea-bb3c-dec0fca28143.png)

**Comment by MelbourneDeveloper** [comment]
One bug question here is why Dart lets me get in to this mess in the first place...

I've all the strictness on I think...

**Comment by dnys1** [comment]
Seems to be due to this: https://github.com/dart-lang/sdk/issues/49419. As for a fix, you could take the approach they recommend of storing the closures as `Function` types but you end up needing to either cast in some places or work exclusively with `Function` which provides no static checks. 

This also seems to work:

```dart
const ServiceDefinition(
  this.factory, {
  this.isSingleton = false,
  void Function(T)? dispose,
}) : _dispose = dispose;

final void Function(T service)? _dispose;

///The dispose method that is called when you dispose the scope
void dispose(T instance) {
  _dispose?.call(instance);
}

// ...

extension IocContainerExtensions on IocContainer {
  ///Dispose all items in the scope. Warning: don't use this on your root
  ///container. You should only use this on scoped containers
  void dispose() => singletons.forEach(
        (key, value) {
          serviceDefinitionsByType[key]!.dispose(value);
        },
      );
}
```

**Comment by MelbourneDeveloper** [comment]
Thankyou @dnys1 I will look into this. Just so I understand this... This is something that the Dart team is aware of, and may choose to address at some point in the future?

**Comment by dnys1** [comment]
From their perspective, it's working as expected. Although I don't quite follow the explanation as to why instance members throw at the time of reading the property whereas instance methods throw at the time of calling, since that is the only difference between my code and yours.

**Comment by MelbourneDeveloper** [comment]
@dnys1 that is exactly how I see it

**Comment by MelbourneDeveloper** [comment]
@dnys1 funnily, this works:

```dart
class ServiceDefinition<T> {
  ///Defines a factory for the service and whether or not it is a singleton.
  const ServiceDefinition(
    this.factory, {
    this.isSingleton = false,
    this.dispose,
  });

  ///If true, only once instance of the service will be created and shared for
  ///for the lifespan of the app
  final bool isSingleton;

  ///The factory that creates the instance of the service and can access other
  ///services in this container
  final T Function(
    IocContainer container,
  ) factory;

  ///The dispose method that is called when you dispose the scope
  //Note: we use dynamic here because using T results in a runtime typing error
  // ignore: avoid_annotating_with_dynamic
  final void Function(T service)? dispose;

  ///Creates a new instance of the service definition as a singleton
  ServiceDefinition<T> asSingleton() => ServiceDefinition<T>(
        factory,
        isSingleton: true,
        dispose: dispose,
      );

  ///doc
  void dispose2(T instance) {
    dispose?.call(instance);
  }
}
```

But, if I move it to an extension, it does not work

```dart
///doc
extension ServiceDefinitionExtension<T> on ServiceDefinition<T> {
  ///doc
  void dispose2(T instance) {
    dispose?.call(instance);
  }
}
```

Something is fishy about that...

**Comment by MelbourneDeveloper** [comment]
@dnys1 thanks to your code I was able to fix the issue here edae98bf117429234b7f3e90fae65ef250b71cd3
